File: app/src/test/java/com/example/myapplication/core/program/SchedulePlannerOverflowTest.kt
package com.example.myapplication.core.program

import org.junit.Assert.assertThrows
import org.junit.Test

class SchedulePlannerOverflowTest {
    @Test
    fun dueDateOverflowIsRejectedInsteadOfWrapping() {
        assertThrows(ArithmeticException::class.java) {
            SchedulePlanner.dueEpochDays(Long.MAX_VALUE, listOf(0, 0))
        }
    }

    @Test
    fun latenessOverflowIsRejectedInsteadOfWrapping() {
        assertThrows(ArithmeticException::class.java) {
            SchedulePlanner.carryForwardAfterCompletion(
                dueEpochDays = listOf(Long.MIN_VALUE, 0L),
                completedIndex = 0,
                completionEpochDay = Long.MAX_VALUE,
            )
        }
    }

    @Test
    fun shiftedDueDateOverflowIsRejectedInsteadOfWrapping() {
        assertThrows(ArithmeticException::class.java) {
            SchedulePlanner.carryForwardAfterCompletion(
                dueEpochDays = listOf(0L, Long.MAX_VALUE),
                completedIndex = 0,
                completionEpochDay = 1L,
            )
        }
    }
}

File: flutter/test/core/program/schedule_planner_overflow_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_app/core/program/schedule_planner.dart';

void main() {
  group('SchedulePlannerOverflowTest', () {
    const int maxInt = 9223372036854775807;
    const int minInt = -9223372036854775808;

    test('dueDateOverflowIsRejectedInsteadOfWrapping', () {
      expect(
        () => SchedulePlanner.dueEpochDaysFromRestDays(
          startEpochDay: maxInt,
          restDays: const [0, 0],
        ),
        throwsA(isA<UnsupportedError>()),
      );
    });

    test('latenessOverflowIsRejectedInsteadOfWrapping', () {
      expect(
        () => SchedulePlanner.carryForwardAfterCompletion(
          dueEpochDays: const [minInt, 0],
          completedIndex: 0,
          completionEpochDay: maxInt,
        ),
        throwsA(isA<UnsupportedError>()),
      );
    });

    test('shiftedDueDateOverflowIsRejectedInsteadOfWrapping', () {
      expect(
        () => SchedulePlanner.carryForwardAfterCompletion(
          dueEpochDays: const [0, maxInt],
          completedIndex: 0,
          completionEpochDay: 1,
        ),
        throwsA(isA<UnsupportedError>()),
      );
    });
  });
}
